internal/provider/newrelic/provider_integration_test.go

```diff
@@ -123,12 +135,205 @@ func Test_Getting_external_metric_generates_a_query_not_rejected_by_backend(t *t
+	t.Parallel()
 
-	testEnv := testutil.TestEnv{}
+	ctx := testutil.ContextWithDeadline(t)
+
+	testEnv := testutil.TestEnv{
+		StartKubernetes: true,
+		Region:          "Local",
+	}
 	testEnv.Generate(t)
 
+	p := newrelicProviderWithMetric(t, newrelic.Metric{
+		Query: testIntegrationQuery,
+	}, &testEnv)
+
+	runAdapter(t, &testEnv, p)
+
+	//nolint:paralleltest
+	t.Run("when_response_error", func(t *testing.T) {
+		cases := map[string]func(w http.ResponseWriter, r *http.Request){
+			"is_200_with_empty_data": func(w http.ResponseWriter, _ *http.Request) {
+				w.WriteHeader(http.StatusOK)
+			},
+			"is_400": func(w http.ResponseWriter, _ *http.Request) {
+				w.WriteHeader(http.StatusBadRequest)
+			},
+			"is_401": func(w http.ResponseWriter, _ *http.Request) {
+				w.WriteHeader(http.StatusUnauthorized)
+			},
+			"is_403": func(w http.ResponseWriter, _ *http.Request) {
+				w.WriteHeader(http.StatusForbidden)
+			},
+			"is_500": func(w http.ResponseWriter, _ *http.Request) {
+				w.WriteHeader(http.StatusInternalServerError)
+			},
+		}
+
+		for testCaseName, handlerFunc := range cases {
+			handlerFunc := handlerFunc
+			//nolint:paralleltest
+			t.Run(testCaseName, func(t *testing.T) {
+				runHTTPTestServer(t, 3000, handlerFunc)
+
+				m := provider.ExternalMetricInfo{Metric: testMetricName}
+
+				if _, err := p.GetExternalMetric(ctx, "", nil, m); err == nil {
+					t.Error("Error expected")
+				}
+
+				// url reflecting the external metrics endpoint for the test_metrics so that later we can simulate the request a hpa
+				// would perform to fetch metrics.
+				url := fmt.Sprintf("%s/apis/external.metrics.k8s.io/v1beta1/namespaces/*/test_metric", testEnv.BaseURL)
+				checkStatusCodeNotOK(ctx, t, &testEnv, url)
+			})
+		}
+	})
+
+	//nolint:paralleltest
+	t.Run("when_returns_200_for_first_requests_and_then_500", func(t *testing.T) {
+		server := runHTTPTestServer(t, 3000, func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write([]byte(`{"data": {
+				"actor": {
+				  "account": {
+					"nrql": {
+					  "results": [{"one": 0.015}]
+					}
+				  }
+				}
+			}}`))
+		})
+
+		m := provider.ExternalMetricInfo{Metric: testMetricName}
+
+		// Make the 3rd request to fail.
+		for i := 0; i < 3; i++ {
+			if i == 2 {
+				server.Config.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.Header().Set("Content-Type", "application/json")
+					w.WriteHeader(http.StatusInternalServerError)
+				})
+				if _, err := p.GetExternalMetric(ctx, "", nil, m); err == nil {
+					t.Error("Error expected")
+				}
+
+				return
+			}
+
+			if _, err := p.GetExternalMetric(ctx, "", nil, m); err != nil {
+				t.Errorf("Unexpected error = %v", err)
+			}
+		}
+
+		url := fmt.Sprintf("%s/apis/external.metrics.k8s.io/v1beta1/namespaces/*/test_metric", testEnv.BaseURL)
+		checkStatusCodeNotOK(ctx, t, &testEnv, url)
```

Comment:
There's something I'm not fully understanding about this test: Why are not we calling `checkStatusCodeNotOK`/`checkStatusCodeOK` in the for loop above? Seems like we're skipping a layer by testing `p.GetExternalMetric` instead. Is there a reason for this, or do we know the implications of skipping it?

Reply:
This is my fault, I forget to call them in the proper places.

Now that you mention it, I considered either including `checkStatusCodeNotOK` in the `testutil` package or doing a refactor on `CheckStatusCodeOK` func to be more generalist but in the end I went for the easiest approach. WDYT?

Reply:
I would rather refator `CheckStatusCodeOk` so you avoid repeating most of the code and get both cases (_Ok_ and _NotOk_ checked in the same way)

In order to keep the change small, you could keep current `CheckStatusCodeOK` signature and extend testutils to include a new function  receiving the 'status code check logic'. Maybe something like this:

```go
func RetryGetRequestAndCheckStatus(
    ctx context.Context, t *testing.T, httpClient *http.Client, url string, failCondition func(statusCode int) bool,
) []byte {
     // current checkStatusCodeOK implementation using `failCondition` argument to check when it should fail
}


func CheckStatusCodeOK(ctx context.Context, t *testing.T, httpClient *http.Client, url string) []byte {
    return RetryGetRequestAndCheckStatus(
        ctx, t, httpClient, url,
        func (statusCode int) bool {
             return statusCode != http.StatusOK
        },
    )
}
```
What do you think?

Reply:
I really like your suggestion, thanks! Added in 2ec3b57e8225b48f182cc05a1eae3a443837fc3f